### ISSUE TITLE
Data field names are coupled with UI

### ISSUE THREAD

**Issue by owlishDeveloper** [taskcluster/taskcluster#631]
The way `DataTable` component is currently set up, the names of the fields in the data and the names of the columns in the table must be the same. `DataTable` takes in the array of strings - which we define, and which later will become the keys of the columns to sort by. We map the names of the columns to the data fields by using `camelCase` function (see my last link) - which removes the spaces from the string we see in the UI.

This seems restrictive and abstraction leak. UI should not be coupled with the data like that.

It's better to have an array of objects like this: `{id: 'workerType', label: 'Name'}`, where the `label` is what will be displayed, and `id` is what corresponds to the data field and what will be used to sort. That way, even if I have a field in my data that is named `workerType`, I have no obligation to name the column `Worker Type` in the UI - I can name it `Name`, or `Your Lovely Worker Types` or absolutely whatever. And that whatever can be changed any time without the need to change graphql code and migrate database

Example of what I am talking about in the `material-ui` documentation: https://material-ui.com/demos/tables/#sorting-amp-selecting and press `Show the source` button

Relevant code to look at:
https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/DataTable/index.jsx#L127-L141 - definition of the header and the sorting label
https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/DataTable/index.jsx#L63 - array of strings comes into the table
https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/AwsProvisionerWorkerTypeTable/index.jsx#L39-L66 - example of the sorting function.
OR
https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/AwsProvisionerWorkerTypeTable/index.jsx#L41 - you can just start reading code from this line

**Comment by helfi92** [comment]
Agreed. Adding a label would make this neater. We currently use a mapping function for that. For example:

https://github.com/taskcluster/taskcluster/blob/a60f57c3215c46022a3f0d6f59834bb9134359c1/ui/src/components/AwsProvisionerWorkerTypeStatus/index.jsx#L122-L131

**Comment by sudipt1999** [comment]
If I am not getting it wrong we are planning to change the type props is being received in the component, 
something to be received as array where each element would look like this

```javascript
{ id: 'ColumnName', label: 'Worker Type' , 'Instance Type': item.instanceType, 'Availability Zones': item.zone, }
```
where id would represent the `column names`.

If that so I would like to try this one then!

**Comment by helfi92** [comment]
I'm not sure I understand but what I can probably say is that we need to change the `headers` prop in `DataTable` from accepting only an array of strings to being able to receive an array of objects of the form `{ label: '<column-name>', id: '<property-name-in-original-object' }`. Does this help get you started?

**Comment by sudipt1999** [comment]
@helfi92  Yeah! Sure will give it a try !

**Comment by helfi92** [comment]
Awesome, many thanks!

**Comment by helfi92** [comment]
Let's also add a `type` (string or number) property to the object so that we can right align columns that are numbers.

**Comment by helfi92** [comment]
@sudipt1999 How is it going?

**Comment by ruchikabgosain** [comment]
Hey @sudipt1999 are you still working on this one?

**Comment by sudipt1999** [comment]
Sorry ! currently I am on my internship so can't work on it, feel free to
take on this !

**Comment by ruchikabgosain** [comment]
AFAIK, we need to change the `headers` prop in `DataTable` from accepting only an array of strings to being able to receive an array of objects of the form `{ label: '<column-name>', id: '<property-name-in-original-object', type: '<to-right-align-columns>' }`. 

Am I right? @helfi92 Is there anythign else i need to know to work upon this?

**Comment by helfi92** [comment]
@ruchikabgosain You have the right idea :)

**Comment by Jo-IE** [comment]
Hi, I'm looking into this issue, and this is my thought process so far on resolving it:

1) passing in an array of objects as stated in the above comments in `/ui/src/components/AwsProvisionerWorkerTypeTable/index.jsx`

2) changing the values in the header `map` function (https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/DataTable/index.jsx#L134-L141) to `header.label`  where necessary as `header` items are now objects and not strings.

Please could I know what I'm missing so far? @djmitche 

Also, I'm a bit confused on what the `label` property in the objects should be. For example, will the `Worker Type` header become `{label: 'Worker Type', id: 'workerType', type: 'String'}` ?

**Comment by djmitche** [comment]
That sounds about right!  Yes, the label should default to the camelCase version of the id.  In some uses of DataTable, we'll want to override that completely, though -- for example, we might want to write `{label: 'Provisioner', id: 'provisionerId'}`.

**Comment by Jo-IE** [comment]
I think I will wait till my previous PR #1403 is merged before making a PR for this issue because I just tried to make a PR and my commit included changes from the previous PR for issue #1403
